### PR TITLE
Jessica and Kate - Time

### PR DESCRIPTION
**Pull request by Catherina87**
# Assignment Submission: OO Ride Share

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | We spent the first day looking at the code and familiarizing ourselves with the code. We took time to see how methods were connected to other between files and classes. We had some difficulty understanding individual, new pieces of code, like "self.class.validate_id" and where it fit in.
What _inheritance_ relations exist between classes? | Trip, Driver, and Passenger classes all inherit from CsvRecord.
What _composition_ relations exist between classes? | TripDispatcher creates several arrays for each class's objects.
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We were forced to make a decision when considering situations where the price of the trip was below $1.65. We were unsure at first whether how to handle situation of whether the driver should then take all the money or not. We decided to charge the 20% still, but with no fee.
Give an example of a _template method_ that you implemented for this assignment | .from_csv was a template method was was used and applied to subclasses.
Give an example of a _nominal test_ that you wrote for this assignment | For driver's total_revenue, we tested what a middle of the road cost would be like, such as a 10 dollar trip.
Give an example of an _edge case test_ that you wrote for this assignment | For driver's total_revenue, we tested an edge case of where the driver has no trips (0).
What is a concept that you gained more clarity on as you worked on this assignment | Inheritance became more clear as we worked on the project. Reading from CSV's also became more clear, because we better understood the structure. Dependencies were also more clear as we worked, such as how all the classes related to TripDispatcher.

### CODE REVIEW

File: lib/trip_dispatcher.rb [hunk 1]
```diff
@@ -19,6 +21,72 @@ def find_passenger(id)
...
+    end
+
+    def select_driver(available_drivers)
+      available_drivers.each do |driver|
+        if driver.trips.length == 0
+          return driver
+        end
+      end
+      
+      available_drivers.each do |driver|
```
Comment (beccaelenzil) [inline lib/trip_dispatcher.rb:44]: Great use of enumerable method `.sort_by!`

**Comment by beccaelenzil** [comment]
# OO Ride Share



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The code demonstrates individual learning about `Time` and the responsibility of `Trip.from_csv`, and uses `Time.parse` in `Trip.from_csv` | ✔️
The code demonstrates breaking out complex logic in helper methods, such as making a helper method in `Trip` to calculate duration | ✔️
There are tests for the nominal cases for the `Passenger#net_expenditures` and `Passenger#total_time_spent` | ✔️
There is at least one edge case test for either `Passenger#net_expenditures` or `Passenger#total_time_spent` testing if the passenger has no trips | ✔️
Practices inheritence. `Driver` inherits from `CsvRecord`, and implements `from_csv` | ✔️
Employs problem-solving and implements `Driver#average_rating` and `Driver#total_revenue` | ✔️
Implements the `TripDispatcher#request_trip`, which creates an instance of `Trip` with a driver and passenger, adds the new trip to `@trips`, and changes the status of the driver | ✔️
Practices composition. In `TripDispatcher#request_trip`, the driver gets connected to the new trip, the passenger gets connected to the new trip | ✔️
Practices git with at least 10 small commits and meaningful commit messages | ✔️

## Testing Requirements

| Testing Requirement | yes/no |
| --- | --- |
There is reasonable test coverage for wave 1, and all wave 1 tests pass | ✔️
There is reasonable test coverage for wave 2, and all wave 2 tests pass | ✔️
Wave 3: Tests in wave 1 and wave 2 explicitly test that only completed trips should be calculated (and ignore in-progress trips) | ✔️
There is reasonable test coverage for `TripDispatcher#request_trip`, and all tests pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 3+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 6+ in Code Review && 2+ in Functional Requirements | 
| Red (Not at Standard) | 0-5 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this project. It is clear that the learning goals around TDD, inheritance, and object composition were met. You did a particularly good implementing thorough tests. Keep up the hard work.
